### PR TITLE
Return host (numpy) instead of device (jax) arrays from observation getters

### PR DESCRIPTION
**Pull request by sbiquard**
Observation getters feed two consumers: the `ObservationReader`, which runs them inside an `io_callback`, and the (single-obs) mapmakers, which builds jax operators directly. Returning jax (device) arrays served the latter but forced a device->host->device round trip at the callback boundary.

Getters now return host numpy arrays (the natural data-loader output) and each device consumer converts at its own boundary:
- the reader pads with `np.pad` so the callback returns host arrays and `io_callback` does a single host->device transfer;
- the mapmaker/templates wrap getter outputs in `jnp.asarray` where they enter operators.

Also fixes a latent bug in the toast scan-mask getters, which did in-place assignment on a jnp array (now np.zeros).

In passing, also delete legacy toast stuff that (I think) no one uses.

### CODE REVIEW

File: src/furax/interfaces/sotodlib/observation.py [hunk 1]
```diff
@@ -376,16 +375,14 @@ def get_boresight_quaternions(self) -> Float[Array, 'samp 4']:
...
         """Returns the quaternion offsets of the detectors"""
-        quats = quaternion.from_xieta_angles(
-            jnp.array(self.data.focal_plane.xi, dtype=jnp.float64),
-            jnp.array(self.data.focal_plane.eta, dtype=jnp.float64),
-            jnp.array(self.data.focal_plane.gamma, dtype=jnp.float64),
-        )
-        return jnp.atleast_2d(quats)
+        # Use so3g's CPU implementation so the result stays on host.
+        fp = self.data.focal_plane
+        quats = so3g.proj.quat.rotation_xieta(fp.xi, fp.eta, fp.gamma)
```
Comment (ArtemBasyrov) [inline src/furax/interfaces/sotodlib/observation.py:385]: Is the output format the same as before here?

File: src/furax/mapmaking/_observation.py [hunk 2]
```diff
@@ -17,18 +20,40 @@
...
+            uid=jax.ShapeDtypeStruct((), dtype=np.uint32),  # type: ignore[arg-type]
+            telescope_uid=jax.ShapeDtypeStruct((), dtype=np.uint32),  # type: ignore[arg-type]
+            detector_uids=jax.ShapeDtypeStruct((n_dets,), dtype=np.uint32),  # type: ignore[arg-type]
+        )
+
+
+def _names_to_uids(names: str | list[str] | np.ndarray) -> UInt32[np.ndarray, ...]:
+    """Converts names to unsigned 32-bit integers using hashing."""
+    # SHA-1 hash truncated to a non-negative 32-bit integer
+    to_int = lambda s: int(sha1(s.encode()).hexdigest(), 16) & 0xEFFFFFFF
```
Comment (ArtemBasyrov) [inline src/furax/mapmaking/_observation.py]: The moved helper keeps & 0xEFFFFFFF but the new comment says "truncated to a non-negative  32-bit integer." 0xEFFFFFFF only clears bit 28 — it's not a truncation and doesn't map cleanly to "non-negative" either (the result is uint32 regardless). This is preexisting behavior moved verbatim, so no functional concern, but if the original intent was a positive 31-bit int it was probably meant to be 0x7FFFFFFF. Worth either fixing or making the comment match the code.

File: src/furax/mapmaking/_observation.py [hunk 3]
```diff
@@ -124,31 +154,33 @@ def get_demodulated_noise_models(self, stokes: ValidStokesType = 'IQU') -> Any:
...
-    def get_timestamps(self) -> Float[Array, ' a']:
+    def get_timestamps(self) -> Float[np.ndarray, ' a']:
         """Returns timestamps (sec) of the samples"""
 
-    def get_elapsed_times(self) -> Float[Array, ' a']:
+    def get_elapsed_times(self) -> Float[np.ndarray, ' a']:
         """Returns time (sec) of the samples since the observation began"""
         timestamps = self.get_timestamps()
-        return timestamps - timestamps[0]
+        return np.asarray(timestamps - timestamps[0])
```
Comment (ArtemBasyrov) [inline src/furax/mapmaking/_observation.py]: isn't it already a numpy array?

**Review by ArtemBasyrov — COMMENTED** [review]
Looks good, but there are some clarifications required

File: src/furax/interfaces/sotodlib/observation.py (hunk 1 above)
Comment (sbiquard) [reply src/furax/interfaces/sotodlib/observation.py:385]: Yes, we use the same quaternion convention as so3g: (1,i,j,k) i.e. scalar first.
See e.g. https://github.com/simonsobs/so3g/blob/master/python/proj/quat.py

File: src/furax/mapmaking/_observation.py (hunk 2 above)
Comment (sbiquard) [reply src/furax/mapmaking/_observation.py]: Good point, I wanted to clarify the comment and exposed an inconsistency! Switched to `0x7FFFFFFF`.

File: src/furax/mapmaking/_observation.py (hunk 3 above)
Comment (sbiquard) [reply src/furax/mapmaking/_observation.py]: Fair, I think this was just working around a mypy complaint, added `type: ignore` instead.

**Review by ArtemBasyrov — DISMISSED** [review]

File: src/furax/mapmaking/mapmaker.py
Comment (Wuhyun) [inline src/furax/mapmaking/mapmaker.py]: Do we now need jnp.asarray here instead?
`jnp.asarray(observation.get_tods(), dtype=config.dtype)`

Comment (Wuhyun) [inline src/furax/mapmaking/mapmaker.py]: Same here:
`jnp.asarray(observation.get_tods(), dtype=config.dtype)`

Comment (Wuhyun) [inline src/furax/mapmaking/mapmaker.py]: `jnp.asarray(observation.get_tods(), dtype=config.dtype)`

Comment (Wuhyun) [inline src/furax/mapmaking/mapmaker.py]: `jnp.asarray(observation.get_tods(), dtype=config.dtype)`

**Review by Wuhyun — DISMISSED** [review]
This makes a lot of sense, thank you!
Just commented on some parts of the single-obs mapmaker that I think need to be changed, but maybe JAX does it automatically?

**Comment by sbiquard** [comment]
Makes sense, thanks!
